src/Collection/Resources/assets/test/controller.test.ts

```diff
@@ -0,0 +1,40 @@
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+'use strict';
+
+import { Application } from '@hotwired/stimulus';
+import { getByTestId } from '@testing-library/dom';
+import { clearDOM, mountDOM } from '@symfony/stimulus-testing';
+import CollectionController from '../src/controller';
+
+const startStimulus = () => {
+    const application = Application.start();
+    application.register('symfony--ux-collection--collection', CollectionController);
+};
+
+/* eslint-disable no-undef */
+describe('CollectionController', () => {
```

Comment:
Sure, I planned to add a Panther test.

Reply:
I don't think we need panther here if we use the testing library. I currently could not get your branch running. But something into this direction should work to test add and remove:

```js
/*
 * This file is part of the Symfony package.
 *
 * (c) Fabien Potencier <fabien@symfony.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

'use strict';

import { Application } from '@hotwired/stimulus';
import {fireEvent, getByTestId, getByText, waitFor} from '@testing-library/dom';
import { clearDOM, mountDOM } from '@symfony/stimulus-testing';
import CollectionController from '../src/controller';

const startStimulus = () => {
    const application = Application.start();
    application.register('symfony--ux-collection--collection', CollectionController);
};

const emptyCollection = '<div data-testid="collection" data-controller="symfony--ux-collection--collection"></div>';

const simpleCollection = '' +
    '    <template id="addButton">' +
    '        <button type="button" class="add-button">Add button</button>' +
    '    </template>' +
    '' +
    '    <template id="deleteButton">' +
    '        <button type="button" class="delete-button">Delete button</button>' +
    '    </template>' +
    '' +
    '<form name="game" method="post">' +
    '   <div id="game">' +
    '       <div>' +
    '           <label class="required">Teams</label>' +
    '           <div id="game_teams" data-prototype="&lt;div class=&quot;team-entry&quot;&gt;&lt;label class=&quot;required&quot;&gt;__name__label__&lt;/label&gt;&lt;div id=&quot;game_teams___name__&quot;&gt;&lt;div&gt;&lt;label for=&quot;game_teams___name___name&quot; class=&quot;required&quot;&gt;Name&lt;/label&gt;&lt;input type=&quot;text&quot; id=&quot;game_teams___name___name&quot; name=&quot;game[teams][__name__][name]&quot; required=&quot;required&quot; data-controller=&quot;test&quot; /&gt;&lt;/div&gt;&lt;/div&gt;&lt;/div&gt;">' +
    '           </div>' +
    '       </div>' +
    '   <div>' +
    '   <button type="submit" id="game_submit" name="game[submit]">Submit</button>' +
    '   </div>' +
    '</form>'

const nestedCollection = '' +
    '    <template id="addButton">' +
    '        <button type="button" class="add-button">Add button</button>' +
    '    </template>' +
    '' +
    '    <template id="deleteButton">' +
    '        <button type="button" class="delete-button">Delete button</button>' +
    '    </template>' +
    '' +
    '<form name="game" method="post">' +
    '   <div id="game">' +
    '       <div>' +
    '           <label class="required">Teams</label>' +
    '           <div id="game_teams" data-prototype="&lt;div&gt;&lt;label class=&quot;required&quot;&gt;__name__label__&lt;/label&gt;&lt;div id=&quot;game_teams___name__&quot;&gt;&lt;div&gt;&lt;label for=&quot;game_teams___name___name&quot; class=&quot;required&quot;&gt;Name&lt;/label&gt;&lt;input type=&quot;text&quot; id=&quot;game_teams___name___name&quot; name=&quot;game[teams][__name__][name]&quot; required=&quot;required&quot; data-controller=&quot;test&quot; /&gt;&lt;/div&gt;&lt;div&gt;&lt;label class=&quot;required&quot;&gt;Players&lt;/label&gt;&lt;div id=&quot;game_teams___name___players&quot; data-prototype=&quot;&amp;lt;div&amp;gt;&amp;lt;label class=&amp;quot;required&amp;quot;&amp;gt;__name__label__&amp;lt;/label&amp;gt;&amp;lt;div id=&amp;quot;game_teams___name___players___name__&amp;quot;&amp;gt;&amp;lt;div&amp;gt;&amp;lt;label for=&amp;quot;game_teams___name___players___name___firstName&amp;quot; class=&amp;quot;required&amp;quot;&amp;gt;First name&amp;lt;/label&amp;gt;&amp;lt;input type=&amp;quot;text&amp;quot; id=&amp;quot;game_teams___name___players___name___firstName&amp;quot; name=&amp;quot;game[teams][__name__][players][__name__][firstName]&amp;quot; required=&amp;quot;required&amp;quot; /&amp;gt;&amp;lt;/div&amp;gt;&amp;lt;div&amp;gt;&amp;lt;label for=&amp;quot;game_teams___name___players___name___lastName&amp;quot; class=&amp;quot;required&amp;quot;&amp;gt;Last name&amp;lt;/label&amp;gt;&amp;lt;input type=&amp;quot;text&amp;quot; id=&amp;quot;game_teams___name___players___name___lastName&amp;quot; name=&amp;quot;game[teams][__name__][players][__name__][lastName]&amp;quot; required=&amp;quot;required&amp;quot; /&amp;gt;&amp;lt;/div&amp;gt;&amp;lt;/div&amp;gt;&amp;lt;/div&amp;gt;&quot;&gt;&lt;/div&gt;&lt;/div&gt;&lt;/div&gt;&lt;/div&gt;">' +
    '           </div>' +
    '       </div>' +
    '   <div>' +
    '   <button type="submit" id="game_submit" name="game[submit]">Submit</button>' +
    '   </div>' +
    '</form>'

/* eslint-disable no-undef */
describe('CollectionController', () => {
    afterEach(() => {
        clearDOM();
    });

    it('connects', async () => {
        const container = mountDOM(emptyCollection);

        startStimulus();

        // smoke test
        expect(getByTestId(container, 'collection')).toHaveAttribute('data-controller', 'symfony--ux-collection--collection');
    });

    it('add a new collection', async () => {
        const container = mountDOM(simpleCollection);

        startStimulus();

        await waitFor(() => getByText(container, 'Add button'));

        fireEvent(
            getByText(container, 'Add button'),
            new MouseEvent('click', {
                bubbles: true,
                cancelable: true,
            }),
        )

        expect(container.querySelector('.entry')).toBe(1);
    });

    it('remove a collection', async () => {
        const container = mountDOM(simpleCollection);

        startStimulus();

        await waitFor(() => getByText(container, 'Add button'));

        fireEvent(
            getByText(container, 'Add button'),
            new MouseEvent('click', {
                bubbles: true,
                cancelable: true,
            }),
        )

        expect(container.querySelector('.entry')).toBe(1);

        await waitFor(() => getByText(container, 'Delete button'));

        expect(container.querySelector('.entry')).toBe(0);
    });
});
```

Reply:
@dunglas could get it running and updated the code snippet. Currently only test for simple add and remove and no tests yet for the nested collection but already created the required html for it.

Reply:
IMHO it's better to create an E2E test here because we want to be sure that the HTML generated by the form component is still compatible with the JS code.

Reply:
The Panther tests would still be worthy as they would ensure that the submitted data is the one expected by the backend code.

Reply:
Make sense specially @stof usecase to make sure that the submitted data is the one we are expecting. Still I think jest tests should also be added. I moved it to a PR: 

https://github.com/dunglas/ux/pull/5